framework/source/class/qx/ui/table/columnmenu/MenuItem.js

```diff
@@ -60,24 +56,57 @@ qx.Class.define("qx.ui.table.columnmenu.MenuItem",
                      });
   },
 
-  members :
+
+
+  /*
+  *****************************************************************************
+     MEMBERS
+  *****************************************************************************
+  */
+
+  events :
   {
     /**
-     * Keep menu in sync with programmatic changes of visibility
-     *
-     * @param value {Boolean}
-     *   New visibility value
-     *
-     * @param old {Boolean}
-     *   Previous visibility value
+     * Dispatched when a column changes visibility state. The event data is a
+     * boolean indicating whether the table column associated with this menu
+     * item is now visible.
      */
-    _applyVisible : function(value, old)
+    changeVisible : "qx.event.type.Data"
+  },
+
+
+
+  /*
+  *****************************************************************************
+     MEMBERS
+  *****************************************************************************
+  */
+
+  members :
+  {
+    __columnVisible : true, // {Boolean}
+
+
+    // Interface implementation
+    setVisible : function(value)
     {
-      // avoid recursion if called from listener on "changeValue" property
-      if (! this.bInListener)
+      if (value !== this.__columnVisible)
       {
-        this.setValue(value);
+        this.__columnVisible = value;
+
+        // avoid recursion if called from listener on "changeValue" property
+        if (! this.bInListener) {
+          this.setValue(value);
+        }
+
+        this.fireDataEvent("changeVisible", value, this.__columnVisible);
```

Comment:
BTW: setting the value after firing the event is also wrong (a call to `getVisible` inside a listener would give wrong results). Still this needs to be fixed :blush: